### PR TITLE
`Literal` types in `inspect`: `getcoroutinestate` and `getgeneratorstate`

### PR DESCRIPTION
**Pull request by sobolevn**
There are other places in `inspect.pyi` where `Literal` types can be applied. I hope to finish it today.

P.S. Is `CoroutineType` correct in this context? Or do we need `typing.Coroutine`? 🤔

### CODE REVIEW

File: stdlib/inspect.pyi [hunk 1]
```diff
@@ -367,7 +367,9 @@ GEN_RUNNING: Literal["GEN_RUNNING"]
 GEN_SUSPENDED: Literal["GEN_SUSPENDED"]
 GEN_CLOSED: Literal["GEN_CLOSED"]
 
-def getgeneratorstate(generator: Generator[Any, Any, Any]) -> Literal["GEN_CREATED", "GEN_RUNNING", "GEN_SUSPENDED", "GEN_CLOSED"]: ...
+def getgeneratorstate(
+    generator: Generator[Any, Any, Any]
+) -> Literal["GEN_CREATED", "GEN_RUNNING", "GEN_SUSPENDED", "GEN_CLOSED"]: ...
```
Comment (AlexWaygood) [inline stdlib/inspect.pyi:372]:
```suggestion
) -> GEN_CREATED | GEN_RUNNING | GEN_SUSPENDED | GEN_CLOSED: ...
```

Comment (sobolevn) [reply stdlib/inspect.pyi:372]: Thanks, let try it!

**Comment by AlexWaygood** [comment]
Wait, mypy is complaining about a union of literal types? Shouldn't that just be resolved into a single literal-type by mypy?

**Comment by sobolevn** [comment]
@AlexWaygood no, mypy does not recognise `CORO_*` as types. It treats them as variables of literal types, not types themself.

**Comment by AlexWaygood** [comment]
> @AlexWaygood no, mypy does not recognise `CORO_*` as types. It treats them as variables of literal types, not types themself.

Oh, of course, my bad.

**Comment by sobolevn** [comment]
It looks like an issue with `anyio`. You should not pass `Generator` to this function:

```python
» python
Python 3.9.9 (main, Dec 21 2021, 11:35:28) 
[Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import inspect
>>> def gen():
...    yield 1
...    yield 2
... 
>>> g = gen()
>>> inspect.getcoroutinestate(g)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/inspect.py", line 1711, in getcoroutinestate
    if coroutine.cr_running:
AttributeError: 'generator' object has no attribute 'cr_running'
```

**Comment by AlexWaygood** [comment]
> P.S. Is `CoroutineType` correct in this context? Or do we need `typing.Coroutine`? 🤔

I would go with `Awaitable` instead of `typing.Coroutine` or `types.CoroutineType`. Neither `typing.Coroutine` or `types.CoroutineType` are protocols, and it's very rare that a class at runtime will explicitly inherit from either. (In fact, inheriting from `types.CoroutineType` is impossible at runtime.) But `typing.Awaitable` is a protocol, so mypy will recognise that any awaitable object can be passed to the function if we use it as the annotation.

**Comment by AlexWaygood** [comment]
> > P.S. Is `CoroutineType` correct in this context? Or do we need `typing.Coroutine`? 🤔
> 
> I would go with `Awaitable` instead of `typing.Coroutine` or `types.CoroutineType`. Neither `typing.Coroutine` or `types.CoroutineType` are protocols, and it's very rare that a class at runtime will explicitly inherit from either. (In fact, inheriting from `types.CoroutineType` is impossible at runtime.) But `typing.Awaitable` is a protocol, so mypy will recognise that any awaitable object can be passed to the function if we use it as the annotation.

Though I don't really get what `typing.Coroutine` is for tbh, since it's not a protocol.

**Comment by Akuli** [comment]
`getgeneratorstate` needs `gi_running` and `gi_frame` attributes, `getcoroutinestate` needs `cr_running` and `cr_frame`. Neither is suitable for `Awaitable`, which has `__await__`. According to a comment in `tests/stubtest_allowlists/py3_common.txt`, these attributes are on `collections.abc.Coroutine` and `collections.abc.Generator`.

**Comment by sobolevn** [comment]
> I would go with Awaitable instead of typing.Coroutine or types.CoroutineType.

I don't think it is a proper type for this case 🤔 
`Awaitable` only specifies `__await__`, but `getcoroutinestate` is defined like so:

```python
    if coroutine.cr_running:
        return CORO_RUNNING
    if coroutine.cr_frame is None:
        return CORO_CLOSED
    if coroutine.cr_frame.f_lasti == -1:
        return CORO_CREATED
    return CORO_SUSPENDED
```

The only type I found with all needed `cr_*` attrs is `CoroutineType`. But, since I almost never use asyncio, I might be wrong: maybe there's something else.

**Comment by Akuli** [comment]
>since I almost never use asyncio, I might be wrong: maybe there's something else.

Same here :)

**Comment by AlexWaygood** [comment]
> > since I almost never use asyncio, I might be wrong: maybe there's something else.
> 
> Same here :)

Oh, me too

**Comment by AlexWaygood** [comment]
I'm experimenting with things

**Comment by AlexWaygood** [comment]
> I'm experimenting with things

My conclusion is that I really don't know how asyncio works at all, but that I'd probably go with `typing.Coroutine`

**Comment by sobolevn** [comment]
> My conclusion is that I really don't know how asyncio works at all

Good one, made me laugh 😆 

> I'd probably go with typing.Coroutine

Ok! 👍 

---

P.S. Just wanted to say Happy New Year to all of you! 🎅 Thanks for what you do for Python typing and community. Working with python's typing team is a pleasure!

**Comment by AlexWaygood** [comment]
> P.S. Just wanted to say Happy New Year to all of you! 🎅

And to you too!

> Working with python's typing team is a pleasure!

Agreed 😀

**Review by Akuli — APPROVED** [review]
Looks reasonable. The code in anyio that mypy_primer found is very dynamic anyway, and should be safe to ignore.

**Pull request merged**